lib/requests/request_executor.rb

```diff
@@ -10,65 +11,195 @@
+        @_node_selector = NodeSelector.new(self, single_node_topology)
+      end
+    end
+
+    def synchronized
+      @self_lock.synchronize do
+        yield
+      end
+    end
+
+    def validate_urls(initial_urls, certificate)
+      clean_urls = Array.new(initial_urls.length)
+      require_https = !certificate.nil?
+      index = 0
+      while index < initial_urls.length
+        url = initial_urls[index]
+        begin
+          URI.parse(url)
+        rescue MalformedURLException => _
+          raise "The url '#{url}' is not valid"
+        end
+        clean_urls[index] = url.gsub(/\/+$/, "")
+        require_https |= url.start_with?("https://")
+        index += 1
+      end
+      return clean_urls unless require_https
+      initial_urls.each do |initial_url|
+        next unless initial_url.starts_with("http://")
+        unless certificate.nil?
+          raise "The url #{initial_url} is using HTTP, but a certificate is specified, which require us to use HTTPS"
+        end
+        raise "The url #{initial_url} is using HTTP, but other urls are using HTTPS, and mixing of HTTP and HTTPS is not allowed."
+      end
+      clean_urls
+    end
+
+    def new_first_topology_update(input_urls:)
+      initial_urls = validate_urls(input_urls, certificate)
+
+      func = lambda do
+        list = {}
+
+        return if initial_urls.any? do |url|
+          begin
+            server_node = ServerNode.new(url, initial_database)
+
+            update_topology_async(node: server_node).value!
+
+            initialize_update_topology_timer
+
+            @_topology_taken_from_node = server_node
+
+            true
+          rescue StandardError => e
+            RavenDB.logger.warn(e)
+            if initial_urls.empty?
+              @_last_known_urls = initial_urls
+              raise RuntimeError.new("Cannot get topology from server: " + url, e)
+            end
+
+            list[url] = e
+
+            false
+          end
+        end
+
+        topology = Topology.new(@_topology_etag)
+
+        topology_nodes = self.topology_nodes
+        if topology_nodes.nil?
+          topology_nodes = initial_urls.map do |url|
+            ServerNode.new(url, initial_database, "!")
+          end
+        end
+
+        topology.nodes = topology_nodes
+
+        @_node_selector = NodeSelector.new(self, topology)
+
+        if !initial_urls.nil? && !initial_urls.empty?
+          initialize_update_topology_timer
+          return
+        end
+
+        @_last_known_urls = initial_urls
+        details = list.map { |key, value| "#{key} -> #{value&.getMessage}" }.join(", ")
+        throw_exceptions(details)
+      end
+
+      Concurrent::Future.execute(&func)
+    end
+
+    def initialize_update_topology_timer
+      return if @_update_topology_timer
+
+      synchronized do
+        next if @_update_topology_timer
+
+        @_update_topology_timer = Concurrent::TimerTask.new(execution_interval: 60, timeout_interval: 60) do
```

Comment:
It should. The only difference I see is that:
a) `Concurrent::TimerTask` doesn't have separate duration arguments for the first execution and period (however they are both set to 1 minute in Raven client so no problem here)
b) `Concurrent::TimerTask` provides timeout support
c) I'm not sure how this `net.ravendb.client.primitives.Timer` handles exceptions. In Ruby any unhandled error should crash the timer thread and notify all observers of the timer.

Reply:
OK, let's just make sure it doesn't blow up anywhere and stop them in `dispose()`. Not sure about the error handling, but those callbacks passed to it  seem to have their error handling implemented, so that nothing goes out. @ml054 Can you confirm please?

Reply:
In java if exception happen it cancels subsequent invocation, so I only submit action in timer handler.